### PR TITLE
Workspace transpilation trace

### PR DESCRIPTION
**Pull request by ftomassetti**
This PR introduces the new Workspace Transpilation Trace, to represent transpilation of n files to m files and not just 1-to-1 transpilations

### CODE REVIEW

File: playground/src/main/kotlin/com/strumenta/kolasu/playground/EMFSupport.kt [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    transpilationTraceEO.setMultipleContainment(
+        "generatedFiles",
+        this.generatedFiles.map { it.toEObject(resource, mapping) }
+    )
+
+    transpilationTraceEO.setMultipleContainment("issues", this.transpilationIssues.map { it.toEObject() })
+    return transpilationTraceEO
+}
+
+fun <S : Node, T : Node> TranspilationTrace<S, T>.saveAsJson(name: String, vararg ePackages: EPackage): String {
```
Comment (alessiostalla) [inline playground/src/main/kotlin/com/strumenta/kolasu/playground/EMFSupport.kt:112]: Don't we have generic methods to do this already?

**Review by alessiostalla — APPROVED** [review]

File: playground/src/main/kotlin/com/strumenta/kolasu/playground/EMFSupport.kt (hunk 1 above)
Comment (ftomassetti) [reply playground/src/main/kotlin/com/strumenta/kolasu/playground/EMFSupport.kt:112]: You are right. I think we are doing that just to invoke our own methods to convert certain classes to EObject, such as:
```
fun <S : Node, T : Node> TranspilationTrace<S, T>.toEObject(resource: Resource): EObject {
    val transpilationTraceEO = makeTranspilationTraceEObject()
    transpilationTraceEO.setStringAttribute("originalCode", this.originalCode)
    val mapping = KolasuToEMFMapping()
    transpilationTraceEO.setSingleContainment("sourceResult", this.sourceResult.toEObject(resource, mapping))
    transpilationTraceEO.setSingleContainment("targetResult", this.targetResult.toEObject(resource, mapping))
    transpilationTraceEO.setStringAttribute("generatedCode", this.generatedCode)
    transpilationTraceEO.setMultipleContainment("issues", this.transpilationIssues.map { it.toEObject() })
    return transpilationTraceEO
}

fun WorkspaceTranspilationTrace.WorkspaceFile.toEObject(resource: Resource, mapping: KolasuToEMFMapping): EObject {
    val workspaceFileEO = makeWorkspaceFileEObject()
    workspaceFileEO.setStringAttribute("path", this.path)
    workspaceFileEO.setStringAttribute("code", this.code)
    workspaceFileEO.setSingleContainment("result", this.result.toEObject(resource, mapping))
    return workspaceFileEO
}
```

We could probably defining some system to define serializers/unserializers and avoiding this duplication. Maybe we can open an issue for that, what do you think?

Comment (alessiostalla) [reply playground/src/main/kotlin/com/strumenta/kolasu/playground/EMFSupport.kt:112]: Sure, makes sense

Comment (ftomassetti) [reply playground/src/main/kotlin/com/strumenta/kolasu/playground/EMFSupport.kt:112]: I created this issue: https://github.com/Strumenta/kolasu/issues/230

**Pull request merged**